library/pool.c

```diff
@@ -31,41 +32,47 @@ struct pool_obj_s {
 struct pool_s {
     LIST_HEAD(objs, pool_obj_s) pool;
     size_t memsize;
-    size_t count;
+    size_t capacity;
     size_t out;
+    bool is_closed;
 
     void (*clear_func)(void *);
 };
 
 pool_t *pool_new(size_t objsize, size_t count, void (*clear_func)(void *)) {
     pool_t *p = calloc(1, sizeof(pool_t));
     p->memsize = objsize;
-    p->count = count;
+    p->capacity = count;
     p->clear_func = clear_func;
     return p;
 }
 
-void pool_destroy(pool_t *p) {
-    while (!LIST_EMPTY(&p->pool)) {
-        struct pool_obj_s *m = LIST_FIRST(&p->pool);
+void pool_destroy(pool_t *pool) {
+    pool->is_closed = true;
+
+    while (!LIST_EMPTY(&pool->pool)) {
+        struct pool_obj_s *m = LIST_FIRST(&pool->pool);
         LIST_REMOVE(m, _next);
         free(m);
     }
 
-    free(p);
+    if (pool->out == 0)
+        free(pool);
 }
 
 bool pool_has_available(pool_t *pool) {
-    return !LIST_EMPTY(&pool->pool) || pool->count > 0;
+    assert(!pool->is_closed);
```

Comment:
that's pretty violent.  better to do WTF log?

Reply:
this flags a programming error, i.e. should never happen